Book.java

```diff
@@ -0,0 +1,315 @@
+
+        // Not sure whether I should be passing the owner...
+        this.owner = owner;
+
+        // A book is initially created with a cash account and an equityAccounts account (it's the simplest possible book)
+        cashAccount = new Account("cash", AccountType.ASSET);
+        equityAccount = new Account("equityAccounts", AccountType.EQUITY);
+        addAccount(cashAccount, null);
+        addAccount(equityAccount, null);
+    }
+
+
+    private Agent owner;
+    private HashSet<Account> accounts;
+    private HashSet<Account> assetAccounts;
+    private HashSet<Account> liabilityAccounts;
+    private HashSet<Account> equityAccounts;
+    private HashMap<Class<? extends contracts.Contract>, Account> assetAccountsMap;
+    private HashMap<Class<? extends contracts.Contract>, Account> liabilityAccountsMap;
+    private Account cashAccount;
+    private Account equityAccount;
+
+    public double getAssetValue() {
+        double assetTotal = 0;
+        for (Account assetAccount : assetAccounts) {
+            assetTotal+=assetAccount.getBalance();
+        }
+        return assetTotal;
+    }
+
+    public double getLiabilityValue() {
+        double liabilityTotal = 0;
+        for (Account liabilityAccount : liabilityAccounts) {
+            liabilityTotal+=liabilityAccount.getBalance();
+        }
+        return liabilityTotal;
+    }
+
+    public double getEquityValue() {
+        double equityTotal = 0;
+        for (Account equityAccount : equityAccounts) {
+            equityTotal += equityAccount.getBalance();
+        }
+        return equityTotal;
+    }
+
+    public double getAssetValueOf(Class<? extends Contract> contractType) {
+        return assetAccountsMap.get(contractType).getBalance();
+    }
+
+    public double getLiabilityValueOf(Class<? extends Contract> contractType) {
+        return liabilityAccountsMap.get(contractType).getBalance();
+    }
+
+    public double getCash() {
+        return cashAccount.getBalance();
+    }
+
+
+    private void addAccount(Account account, Class<? extends Contract> contractType) {
+        accounts.add(account);
+        switch(account.getAccountType()) {
+
+            case ASSET:
+                assetAccounts.add(account);
+                assetAccountsMap.put(contractType, account);
+                break;
+
+            case LIABILITY:
+                liabilityAccounts.add(account);
+                liabilityAccountsMap.put(contractType, account);
+                break;
+
+            case EQUITY:
+                equityAccounts.add(account);
+
+            // Not sure what to do with INCOME, EXPENSES
+        }
+
+    }
+
+    /**
+     * Adding an asset means debiting the account relevant to that type of contract
+     * and crediting equity.
+     * @param contract an Asset contract to add
+     */
+    public void addAsset(Contract contract) {
+        Account assetAccount = assetAccountsMap.get(contract.getClass());
```

Comment:
This will check the lowest level of contract (i.e. not the contract's super class), and create an account of that type. Could an account be of a super class of a Contract?